### PR TITLE
Add faithful C++ API

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44158 VariableKernel calls into scattered C++ api
* **#44087 Add faithful C++ API**

Each op taking a TensorOptions argument now has an additional overload in the C++ frontend where it takes scattered ScalarType, Layout, Device, bool instead of one TensorOptions argument.

If it is a c10-full op, then the scattered version calls into the dispatcher and the gathered version is a proxy calling into the scattered version.
If it is a non-c10-full op, then the gathered version calls into the dispatcher and the scattered version is a proxy calling into the gathered version.

This should minimize the amount of gathering and scattering needed.

This PR is also a prerequisite to remove the re-gathering of arguments that is currently happening in VariableKernel. Currently, VariableKernels gather arguments into a TensorOptions object
to call into the C++ API. In a PR stacked on top of this, VariableKernel will just directly call into the scattered C++ API introduced here and avoid the gathering step.

Differential Revision: [D23492188](https://our.internmc.facebook.com/intern/diff/D23492188/)

### CODE REVIEW

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:168]: If you're going to change the naming convention here, you should also be changing it in the other api modules, as they all apply consistent naming for how to get to Argument form. But let me try to defend the existing naming convention and see if I can convince you that it is good :)

The current convention is:
* Each API module's functions are implicitly assumed to be operating on the data types of that module. So therefore. cpp.Argument, not cpp.to_cpp_argument (the second cpp is redundant).
* API modules are always accessed using qualified accessor (`from tools.codegen.api import cpp`, never `from tools.codegen.api.cpp import *`)
* Types are always fully qualified within the name (CppArgument, not cpp.Argument)
* If you refer to a type that is not in this module, you have to add it to the function name (e.g., `legacydispatcher_exprs`)

In the first version of this code I explicitly qualified names with things like `cpp_argument`, but it was really wordy, which is why I settled on actually using the namespacing mechanism here.

File: tools/codegen/api/dispatcher.py
Comment (ezyang) [inline tools/codegen/api/dispatcher.py:77]: Not sure if I agree with the general API design here, but at the very least, `process_tensoroptions` should be an Enum, not a freeform string.

**Comment by ezyang** [comment]
I think my high level code organization feedback is that the output from `api.cpp` should be more structured.  Instead of having two separate functions which possibly return the same thing and then testing for equality, you should instead have a single function that returns a struct that describes the entire structure of the C++ bindings for this function. This struct would contain a mandatory "schema style" signature, as well as an optional TensorOptions convenience signature.

Furthermore, when binding to the dispatcher, it would be better to have things be more compositional. The chain of transformations that I'd expect to see in the terminal state is TensorOptions signature -> Schema style signature -> Dispatch call. So instead of processing TensorOptions directly, you should get schema-style (cpp) expressions from the tensor options, and then shunt everything through the same processing path. Your current code does an "optimization" of short circuiting the conversion to schema-style when the underlying dispatcher is still legacy, but if you really insist on this optimization, it should be organized in an easy to delete fashion, since we definitely don't want this long term.

File: tools/codegen/api/cpp.py
Comment (bhosmer) [reply tools/codegen/api/cpp.py:259]: Agree - enlarging on my earlier argument about keeping `group_arguments` (written before I got here :P), I think it'd be better to keep `arguments` and plumb a bool `gather` flag through.

File: tools/codegen/api/dispatcher.py
Comment (bhosmer) [reply tools/codegen/api/dispatcher.py:77]: I think just an optional bool `gather` should do it. (gather rather than scatter because gather is the special case that's going away.)

File: tools/codegen/api/dispatcher.py
Comment (bhosmer) [inline tools/codegen/api/dispatcher.py:90]: It's inconsistent that we assert on this but not the same for `tensor_options`, I'd do both or neither. (Actually, best would be to bundle them into a little named tuple - correct by construction!)

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:374]: This new block should be factored into helper functions, so the logic of the mainline code stays easy to read. Also, calling the helpers from their actual points of use will make it clearer who depends on what.

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:377]: Hmmm there's definitely a better way to test for the presence/absense of tensor options :P

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:392]: ditto helpers

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:394]: ditto test

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:464]: ditto helpers

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:481]: ditto test

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:496]: ditto helpers

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:498]: ditto test

**Review by bhosmer — COMMENTED** [review]
Couple of API suggestions, plus there's really a lot of new code that I'd love to see factored into helpers, for readability, robustness etc.

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:259]: ok I changed it to a flag

File: tools/codegen/api/dispatcher.py
Comment (smessmer) [reply tools/codegen/api/dispatcher.py:77]: I don't think an optional bool would cut it here. It's hard to understand that `gather=None` means "do nothing" but `gather=False` means "scatter"...I made it an enum

File: tools/codegen/api/dispatcher.py
Comment (smessmer) [reply tools/codegen/api/dispatcher.py]: this assertion is just to make the switch/case comprehensive and make sure we didn't miss a case. The `tensor_options` argument is mostly independent from that.

**Comment by smessmer** [comment]
Major rewrite based on feedback. PTAL

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:199]: If any of the other `tools.codegen.api` modules want to make use of these types it will be much easier to put it in `tools.codegen.api.types` as you will be able to avoid import cycles this way.

File: tools/codegen/api/dispatcher.py
Comment (ezyang) [inline tools/codegen/api/dispatcher.py:79]: nit: I'm personally more a fan of the `ProcessTensoroptions = Enum('ProcessTensoroptions', ('GATHER', 'SCATTER', 'PASS_THROUGH'))` syntax as the numbers get automatically allocated this way. (Also, did you mean TensorOptions, capital-O?)

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:200]: Make this Tuple instead of Sequence, then this dataclass will be hashable.

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:211]: nit: I probably wouldn't enshrine this into the data model (I understand this is refactored from some code that used to live in gen.) The reasoning is that with `exclude_this`, I now have to be aware if a `CppSignature` that I have in hand is either a full CppSignature (with everything) or a "stripped" version (post calling `exclude_this`). But you always *immediately* call some sort of rendering function (e.g., `cpp_arguments_str`) afterwards. We should make the illegal state (a free floating stripped `CppSignature` object) unrepresentable, and we can do this by pushing the defaulting and exclude this behavior directly into kwargs for the relevant rendering functions. WDYT?

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:217]: Reasoning similar to `exclude_this` applies here

File: tools/codegen/api/cpp.py
Comment (ezyang) [inline tools/codegen/api/cpp.py:256]: nit: naming. Binding to me implies "we are binding some arguments into names". But this class doesn't do any binding at all, it's just a collection of signatures. Call it CppSignatures or CppSignatureGroup.

File: tools/codegen/gen.py
Comment (ezyang) [inline tools/codegen/gen.py:324]: It's almost definitely worth a comment explaining why without a gathered signature you don't need to strip defaults, but with a gathered signature you DO need to strip defaults.

File: tools/codegen/gen.py
Comment (ezyang) [inline tools/codegen/gen.py:345]: I'm not sure why this proxying business is going on here. Wouldn't it be simpler to just define translation from "signature to dispatcher" and from "gathered signature to dispatcher" and call it a day? Then you would have only a *single* calling template to populate (the dispatcher code) and you can dedupe the four variants below. I agree that for the signatures above you still need to special case because of defaulting behavior, but for the inside implementations of these functions, I'm pretty sure the template should be the same in all situations.

Comment (ezyang) [reply tools/codegen/gen.py:345]: Let me try to elucidate the underlying design constraint here. I don't think there is anything fundamentally wrong with proxying gathered signature to scattered signature. But it introduces a ton of new translations:

This PR adds:
* Scattered cpp -> gathered cpp translation
* Gathered cpp -> scattered cpp translation
* Gathered cpp -> dispatcher translation
* Scattered cpp -> dispatcher translation

That's a ton of stuff! It seems conceptually simpler to me to reduce the number of translations you need:

* Scattered cpp -> dispatcher translation
* Gathered cpp -> dispatcher translation

You don't need the scattered cpp to gathered cpp translation, because you can just go *directly* to dispatcher. It also makes it harder to make mistakes, as you don't have scattered/gather cpp in the covariant return position, so you don't have to worry about mixing the two of them up.

File: tools/codegen/api/cpp.py
Comment (ezyang) [reply tools/codegen/api/cpp.py:211]: Though, it turns out I may have more fundamental problems with this class. Check https://github.com/pytorch/pytorch/pull/44087/files#r496137831

File: tools/codegen/gen.py
Comment (smessmer) [reply tools/codegen/gen.py:345]: this doesn't work unfortunately. Even if you go directly to the dispatcher, you'll need all 4 translations. There's 2 C++ API calling conventions and 2 dispatcher calling conventions and you need all 4 edges between those nodes.

File: tools/codegen/gen.py
Comment (smessmer) [reply tools/codegen/gen.py:370]: I made `exclude_this` and `with_defaults` into parameters instead of functions. I know this only partially addresses your issue but it's the best I came up with after several attempts.

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:256]: I renamed it to `CppSignatureGroup`.

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:217]: I made it into a parameter

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:211]: I made it into a parameter

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:200]: A tuple has a fixed number of elements right? How does this work for a variable length argument list?

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:199]: I don't agree. Yes, putting everything into one module avoids import cycles, but it also makes things much less structured. Note also that `CppSignature` here is not actually part of the model. It's something computed from the model that only lives temporarily while the corresponding operator is generated.

File: tools/codegen/api/cpp.py
Comment (ezyang) [reply tools/codegen/api/cpp.py:200]: `Tuple[Union[Argument, TensorOptionsArguments, ThisArgument], ...]`

**Comment by ezyang** [comment]
I spent some time working through this diff and trying out some alternative designs, and I think I have a better understanding for which way I'd like the diff to go. I'm hoping the discussion here will help get us more aligned on future new codegen changes.

**Don't be afraid to define single-use NamedTuples for very localized situations.** In this diff, we have to make a determination from a FunctionSchema whether or not we will need two C++ signatures or one. This determination currently lives `signature_group` (previously `group_arguments`) when we scan through the arguments in the schema and work out whether or not there is a set of arguments that group into TensorOptions. So, it seems you worked out that the easiest way to add this functionality without having to reimplement the scanning logic (a good call, by the way) is to modify the signature of `signature_group` to return a new structure  `CppBinding` (as it was called at the time), that recorded both the gathered and ungathered versions.

The miscommunication that happened during review was that I misunderstood the intent of this struct. What *you* needed was a struct to report results from `signature_group`; what *I* thought the struct was supposed to be was a new abstraction for representing CPP API signatures, similar to how we already had abstractions for CppExpr and CppArg. This means that when I requested certain things (such as recording return types, and being able to render the entirety of the signature) which would make sense for a C++ abstraction, these requests would have lead to awkward implementation, as `signature_group` would have ended up having to handle a bunch of extra things that didn't really have anything to do with grouping arguments.

When I got to the same position in my variant, I made a *separate* struct for representing the result of `gather_arguments` (`signature_group` in your PR):

```
class GatherArgumentsResult(NamedTuple):
    arguments: Sequence[Union[Argument, ThisArgument], ...]
    gathered_arguments: Optional[Sequence[Union[Argument, TensorOptionsArguments, ThisArgument], ...]]
```

modified `gather_arguments` to return this struct:

```
def gather_arguments(
    func: FunctionSchema, *, method: bool
) -> GatherArgumentsResult
```

and then maintained a separate struct hierarchy to represent the actual C++ signature:

```
class CppSignature:
    # The schema this signature is derived from
    func: FunctionSchema
    # If this signature is a method, this is not None and contains
    # the corresponding ThisArgument for this signature
    method: Optional[ThisArgument]

    cpp_arguments: Tuple[CppArgument, ...]
    cpp_return_type: str

@dataclass(frozen=True)
class CppSignatureGroup:
    func: FunctionSchema
    signature: CppSignature
    gathered_signature: Optional[CppSignature]
```

No rendering goes on `GatherArgumentsResult`; it's just a vehicle for `gather_arguments` to return its results, and the constructor for `CppSignatureGroup` is what is actually responsible for assembling them into a full-blown concept of a C++ signature. I think this should also resolve your single-responsibility concerns, and it is a lot less weird to store, e.g., the return type in a `CppSignature`, than it is to store it in  `GatherArgumentsResult` (where it makes no semantic sense.)

Arguably, this is the same as what you had before, minus an extra CppSignature struct. But I think having a CppSignature is definitely worth it. In the old codegen, we manually pull out arguments and returns, and it is already a bit wordy; and as you saw in the first version of your patch, it gets *even* more wordy when you have two versions of the C++ signature running around. Encapsulating them into a single struct makes sense, and is a refactor that could have been done prior to doing this patch.

**CppArgument should accurately reflect the C++ arguments of your function; if they don't, you extend the concept of CppArgument to model groups.** Something that seemed a bit smelly to me in early review was the new enum for `SCATTER` and `GATHER` and `PASS_THROUGH`. It seemed to me that this shouldn't be necessary, and I tried to outline the high level design reasoning in https://github.com/pytorch/pytorch/pull/44087#discussion_r496133606 although you found it difficult to actually solve this problem entirely. I think I know why.

In the old codegen, there is a big coincidence: when mapping from cpp to dispatcher, it is always the case that a cpp argument maps to one or more dispatcher arguments, but *never vice versa* (never mapping multiple cpp arguments to one dispatcher argument). In your faithful C++ API world this is no longer the case: you may have to map a scattered tensor options (from the faithful API) into a single dispatcher argument (non-faithful API). This is very inconvenient to do with the CppArgument abstraction, because if you actually had separate CppArguments for each scattered argument, then you have to somehow group them back together to map them to a single TensorOptions dispatcher argument. You work around this in your PR by always passing in the *gathered* CppArguments, and then using `process_tensoroptions` to disambiguate whether or, no, actually, the binding sites were scattered CppArguments.

The old saying is that all problems in computer science can be solved with another level of indirection. I think this applies here too. If you need CppArguments to be grouped together, *add the concept of a group of CppArguments*. It would look something like this, in analogy to the existing `TensorOptionsArguments` struct:

```
@dataclass(frozen=True)
class CppArgumentGroup:
    def arguments() -> Sequence[CppArgument]:
        raise NotImplementedError

@dataclass(frozen=True)
class CppTensorOptionsArguments:
    argument: TensorOptionsArguments
    dtype: CppArgument
    layout: CppArgument
    device: CppArgument
    pin_memory: CppArgument
    def arguments(self):
        return [self.dtype, self.layout, self.device, self.pin_memory]
```

Most functions that operate on `CppArgument` today would now operate on `Union[CppArgument, CppArgumentGroup]`. In cases where the grouping information is not necessary (e.g., when rendering the declaration/definition signatures in C++), we simply expand the groups into a single CppArgument list. However, when the grouping is necessary (e.g., when computing dispatcher expressions), you then just directly operate on CppArgumentGroup, and your translation problems melt away: you don't need a scatter/gather argument, because you can work out the input convention from the CppArguments you are given as input, and the output convention via is `use_c10_dispatcher: full`.

This makes it hard to make a mistake, because you always feed in the *precise* description of what arguments are in scope to expressions (based on what you actually used to make the signature), and that description is what drives the expression process, not some side-band information

**The prototype.** I pushed an illustrative sketch of some of these ideas to https://github.com/ezyang/pytorch/commit/5c05730a821b06273d6ce495a6262060b1e1f63a It doesn't implement CppArgumentGroup, I haven't even type checked it, but I thought relatively carefully about how things should be put together and it lead me to some new observations about of your code which I didn't notice in CR. Let me know what you think.

Also, I'm putting the effort in here to communicate design ideas because I hope you'll be making more changes to the new codegen in the future. If you think that it makes more sense to unblock you by having me finish up this version of the change, or us agreeing to disagree and landing a basic version to be refactored later, please let me know.

**Comment by smessmer** [comment]
I'm ok with that design. Can you rebase your changes on top of my PR and land them separately? I've already done 3 different versions of this PR based on your feedback and would rather you implemented your vision in a refactoring PR on top instead of me doing a fourth version that again isn't 100% certain to be what you expected.

**Review by ezyang — APPROVED** [review]
@smessmer and I agreed to let me take a whack at refactoring this after landing.

**Pull request closed without merging**
